File: calculator.py
def hasPriority(op1, op2):
    print("op1 "+ op1)
    print("op2 " + op2)
    """"""
    if op1 == '(' or op1 == ')':
        return False
    if (op1 == '*' or op1 == '/') and (op2 == '+' or op2 == '-'):
        return False

    return True


def operations(op, a, b):
    """"""
    if op == '+':
        return a+b
    elif op == '-':
        return a-b
    elif op == '*':
        return a*b
    elif op == '/':
        if b == 0:
            print("cannot divided by 0")
        else:
            return a/b
    elif op == '^':
        return pow(a, b)
    else:
        print("Invalid Operation")


def calculate(string):
    """"""
    operators = ['+', '-', '*', '/', '^']
    values = []
    ops = []
    i=0
    while i < len(string):
        if string[i].isdigit():
            buffer = ''
            print("buffer 1: "+ str(buffer))
            while(string[i].isdigit()):
                buffer += string[i]
                i+=1
            i-=1
            print("buffer "+ str(buffer))
            print("look at i "+ str(i))
            values.append(buffer)
        elif string[i] == '(':
            ops.append(string[i]) 
        elif string[i] == ')':
            while(ops and hasPriority(string[i], ops[-1])):
                values.append(operations(ops.pop(), values.pop(), values.pop()))
            ops.pop()
        elif string[i] in operators:
            print(string[i])
            while(ops and hasPriority(string[i], ops[-1])):
                values.append(operations(string[i], values.pop(), values.pop()))
                print("look: "+ str(values))
            ops.append(string[i])
        i+=1

    while ops:
        values.append(operations(ops.pop(), values.pop(), values.pop()))

    return values.pop() or None

print(calculate('412/4+(3+4/2+5)'))

File: calculator.js
const calculate = s => {
    // Stack for numbers: 'values'
    s = s.trim(' ')
    const values = []
    // Stack for Operators: 'ops'
    const ops = []
  
    for (let i = 0; i < s.length; i++) {
      // Current token is a number, push it to stack for numbers
      if (s[i] >= '0' && s[i] <= '9') {
        let buffer = ''
        // There may be more than one digits in number
        while (i < s.length && (s[i] >= '0' && s[i] <= '9') || s[i]  == '.') {
          buffer += s[i++]
        }
        values.push(Number(buffer))
        i-- 
      } else if (s[i] === '(') { // Current token is an opening brace, push it to 'ops'
        ops.push(s[i])
      } else if (s[i] === ')') {
        while (ops[ops.length - 1] !== '(') { // Closing brace encountered, solve entire brace
          values.push(applyOp(ops.pop(), values.pop(), values.pop()))
        }
        ops.pop()
      } else if (s[i] === '+' || s[i] === '-' || s[i] === '*' || s[i] === '/') { // Current token is an operator.
        // While top of 'ops' has same or greater precedence to current
        // token, which is an operator. Apply operator on top of 'ops'
        // to top two elements in values stack
        while (ops.length && hasPrecedence(s[i], ops[ops.length - 1])) {
          values.push(applyOp(ops.pop(), values.pop(), values.pop()))
        }
        // Push current token to 'ops'.
        ops.push(s[i])
      }
    }
  
    // Entire expression has been parsed at this point, apply remaining
    // ops to remaining values
    while (ops.length) {
      values.push(applyOp(ops.pop(), values.pop(), values.pop()))
    }
  
    // Top of 'values' contains result, return it
    return values.pop()
}
  
  // Returns true if 'op2' has higher or same precedence as 'op1',
  // otherwise returns false.
const hasPrecedence = (op1, op2) => {
  if (op2 === '(' || op2 === ')') {
    return false
  }
  if ((op1 === '*' || op1 === '/') && (op2 === '+' || op2 === '-')) {
    return false
  }
  return true
}
  
  // A utility method to apply an operator 'op' on operands 'a'
  // and 'b'. Return the result.
const applyOp = (op, b, a) => {
  switch (op) {
    case '+':
      return a + b
    case '-':
      return a - b
    case '*':
      return a * b
    case '/':
      if (b === 0) {
        throw 'Cannot divide by zero'
      }
    return Math.floor(a / b)
  }
  return 0
}

console.log(calculate('123.2+2*1+3+2/2'))
console.log(calculate('4/2+(2+2/2+2)'))
